gemini/genewise.py

```diff
@@ -0,0 +1,109 @@
+                "gt_phred_ll_homalt"]
+    return [x for x in all_cols if x in gt_filter and not x in cols]
+
+
+def gen_results(rows, gt_filters, min_filters, min_variants, columns):
+    # we track the index of the passed filter in passed_filters.
+    gene_passed_filters = {}
+    subset = []
+    for row in rows:
+        cols = {c: row[c] for c in columns}
+        # have to test all filters since 1 row can meet multiple fitlers.
+        row_passed_filters = []
+        for i, gt_filter in enumerate(gt_filters, start=1):
+            if eval(gt_filter, cols):
+                # track that this filter passed.
+                gene_passed_filters[i] = True
+                row_passed_filters.append(i)
+        if row_passed_filters:
+            row.print_fields['variant_filters'] = ",".join(map(str, row_passed_filters))
+            subset.append(row)
+    if len(gene_passed_filters) < min_filters or len(subset) < min_variants:
+        raise StopIteration
+
+    # e.g. 1,2 indicating which filters passed
+    passed_filters = ",".join(str(x) for x in sorted(gene_passed_filters))
+    for row in subset:
+        row.print_fields['n_gene_variants'] = len(subset)
+        row.print_fields['gene_filters'] = passed_filters
+        yield row
+
+
+def genewise(db, gt_filters, filter=None, columns=None, min_filters=None,
+             min_variants=1,
+             grouper="gene"):
+    assert os.path.exists(db)
+
+    orig_columns = [x.strip() for x in (columns or "").split(",")]
+    added_cols = add_cols(orig_columns, "||".join(gt_filters))
+    if grouper not in orig_columns:
+        added_cols.append(grouper)
+    columns = orig_columns + added_cols
+    assert not any(';' in c for c in columns)
+
+    # NOTE: we could make the WHERE part customizable.
+    query = "SELECT {columns} FROM variants WHERE (is_exonic = 1 AND impact_severity != 'LOW')"
+    if filter:
+        query += " AND  " + filter
+    query += " ORDER BY CHROM, gene"
+
+    gq = GeminiQuery(db, include_gt_cols=True)
+
+    # use the GeminiQuery machinery to transform to something that's eval'able
+    # then compile once for speed.
+    cleaned_filters = []
+    for gt_filter in gt_filters:
+        assert gq._is_gt_filter_safe(gt_filter)
+        gt_filter = gq._correct_genotype_filter(gt_filter)
+        cleaned_filters.append(compile(gt_filter, gt_filter, 'eval'))
+
+    gq.run(query.format(columns=", ".join(columns)))
+
+    if isinstance(grouper, basestring):
+        grouper = operator.itemgetter(grouper)
+
+    header_printed = False
+    for groupkey, grp in it.groupby(gq, grouper):
+        grp = list(grp)
+        for x in gen_results(list(grp), cleaned_filters, min_filters or 0,
+                             min_variants, columns):
+            for c in added_cols:
+                if c != 'gene':
+                    del x.print_fields[c]
+            if not header_printed:
+                print "\t".join(x.print_fields.keys())
```

Comment:
these are printing the result of the evaluation

I'll add the image to the example that you pointed out above. Thanks!